File: sudoku.rb
#!/usr/bin/env ruby

#index ranges for rows/columns. 
$PUZZLE_RANGE = (0..8).to_a

#puzzle type accomodates for different potential charsets
$TYPE = 0

#reads input file of unsolved sudoku puzzle, saves in 2D array of integers, denotes blank cells as 0
def readPuzzle(file)
    grid = Array.new(9) { Array.new(9) }
    input = Array.new

    f = File.read(file).split("")
    f.each do |a|
        if a.match(/\./)
            input << "."
        elsif a.match(/\d/)
            input << a.to_i
            $TYPE = 1
        elsif a.match(/[A-I]/)
            input << a
            $TYPE = 2
        elsif a.match(/[a-i]/)
            input << a
            $TYPE = 3
        end
    end

    $PUZZLE_RANGE.each do |a|
        grid[a] = input.slice(0,9)
        input.shift(9)
    end
  
    return grid
end

def setTokens()
    if $TYPE == 1
        return (1..9).to_a
    elsif $TYPE == 2
        return ("A".."I").to_a
    elsif $TYPE == 3
        return ("a".."i").to_a
    end
end

def possibleValues(grid, x, y) 
    boxVals = []
  
    tokens = setTokens()

    rowVals = 9.times.collect { |col| grid[x][col] }
    colVals = 9.times.collect { |row| grid[row][y] }

    #checks for tokens in 3x3 box
    boxX = x / 3
    boxY = y / 3

    ((boxX * 3)..((boxX * 3) + 2)).each do |i|
        ((boxY * 3)..((boxY * 3) + 2)).each do |j|
            boxVals << grid[i][j]
        end
    end
  
    tokens -= rowVals.uniq
    tokens -= colVals.uniq	
    tokens -= boxVals.uniq
end
       
#driving solve function
def solve(grid)
    $PUZZLE_RANGE.each do |i|
        $PUZZLE_RANGE.each do |j|
            if grid[i][j] == "."
	        temp = grid.dup

	    tokens = possibleValues(grid, i, j)
	    if tokens.empty?
	        return nil
	    end
				
	    tokens.each do |token|
	        temp[i][j] = token
	        if solve(temp)
	            return temp
	    end
	    temp[i][j] = "."
	  end
          return nil
        end
     end
  end
  return grid
end
				

#janky sudoku print
def prettyPrint(grid)
    print " ----------------------- \n"
    $PUZZLE_RANGE.each do |i|
        if i % 3 == 0 and i != 0
           print "|-------+-------+-------|\n"
        end
        print "| "

        $PUZZLE_RANGE.each do |j|
        if j % 3 == 0 and j != 0
            print "| "
        end
        print "#{grid[i][j]} "
      end
    print "|\n"
    end
print " ----------------------- \n"
end

#check for an input argument, default to reading "puzzle"
if ARGV.empty?
    file = "puzzle"
else
    file = ARGV[0]
end

#read, solve, print
grid = readPuzzle(file)
prettyPrint(grid)
print "\nSolving puzzle...\n\n"
solution = solve(grid)
prettyPrint(solution)

File: sudoku.py
import re
import string
import sys

PUZZLE_RANGE = [*range(0,9,1)]  # index ranges for individual rows/columns.
CELL_NUMBER = 81                # number of cells
HEIGHT = 9                      # height of grid
WIDTH = 9                       # width of grid


# take 81-char string representing sudoku puzzle, transform into 9x9 grid
# determine appropriate character set based on input
def parse_puzzle(input):
    print("Reading puzzle...")
    char_type = 1
    grid = [[0] * HEIGHT] * WIDTH
    arr = []

    f = list(input)
    for a in f:
        if re.match(r'\.', a):
            arr.append(a)
        elif re.match(r'\d', a):
            char_type = 1
            arr.append(int(a))
        elif re.match(r'[A-I]', a):
            char_type = 2    
            arr.append(a)
        elif re.match(r'[a-i]', a):
            char_type = 3     
            arr.append(a)

    for i in PUZZLE_RANGE:
        grid[i] = arr[i*9:i*9+9]
    
    return grid, char_type


# based on input string, assign possible values for individual squares on grid
def set_tokens(char_type):
    if char_type == 1:
        return [*range(1,10,1)]
    elif char_type == 2:
        return list(string.ascii_uppercase[:9])
    elif char_type == 3:
        return list(string.ascii_lowercase[:9])


# determine possible values for a given square based on basic game logic
# exclude values in same row, column and neighboring 3x3 grid
def possible_values(char_type, grid, i, j):
    tokens = set_tokens(char_type)
    
    col_vals, row_vals, box_vals = [], [], []

    for y in range(HEIGHT):
        if grid[y][j] != ".":
            col_vals.append(grid[y][j])

    for x in range(WIDTH):
        if grid[i][x] != ".":
            row_vals.append(grid[i][x])

    boxX = i//3
    boxY = j//3

    for x in range(boxX*3,boxX*3+3):
        for y in range(boxY*3,boxY*3+3):
            if grid[x][y] != ".":
                box_vals.append(grid[x][y])

    tokens = [item for item in tokens if item not in col_vals]
    tokens = [item for item in tokens if item not in row_vals]
    tokens = [item for item in tokens if item not in box_vals]

    return tokens


# recursive backtracking solving algorithm
# i don't feel like explaining it in a comment
def solve(char_type, grid):
    tokens = []
    temp = [[0] * HEIGHT] * WIDTH

    for i in PUZZLE_RANGE:
        for j in PUZZLE_RANGE:
            if grid[i][j] == '.':
                temp = grid
                tokens = possible_values(char_type, grid, i, j)
                if not tokens:
                    return None

                for a in tokens:
                    temp[i][j] = a
                    if solve(char_type, temp):
                        return temp
                    temp[i][j] = '.'
                return None              
    return grid


# command-line printing of a sudoku board of indescribable beauty
def pretty_print(grid):
    print (" ----------------------- ")
    for i in PUZZLE_RANGE:
        if i % 3 == 0 and i != 0:
            print("|-------+-------+-------|")
        print("| ", end='')

        for j in PUZZLE_RANGE:
            if j % 3 == 0 and j != 0:
                print("| ", end='')
            print(str(grid[i][j]) + " ", end='')
        print("|")
    print(" ----------------------- ")


# driving main function.  read puzzle, print, solve, print
def main():
    if len(sys.argv) == 1 or len(sys.argv[1]) != CELL_NUMBER:
        print("Error getting puzzle information")
        quit()

    grid, char_type = parse_puzzle(sys.argv[-1])
    pretty_print(grid)
    print("\nSolving puzzle...")
    solution = solve(char_type, grid)
    pretty_print(solution)


# python name-checking bullshit
if __name__=="__main__":
    main()